Make SparkSqlHook use Connection

uranusjr opened the pull request:
Fix #8713. This takes a different approach from previous work (#12710); instead of deprecating anything, the configuration is now resolved in a two-step process:

1. If an argument is explicitly passed in by the user, it is used (the current ebhaviour), and configuration on the Conenction object is ignored. A try block is added so the Connection is not needed if the user wants to explicitly pass in everything.
2. If the user does not pass in configuration arguments explicitly, the configuration is populated by a Connection object (if available).

Only two options are pulled from the Connection: `master` (to the Connection’s `host` and `port`), and `yarn_queue` (the Connection’s extra value `queue`). Those are also used by `SparkSubmitOperator` and already documented.


---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

uranusjr committed:
Make SparkSqlHook use Connection

This allows a SparkSqlHook to be created without a backing Connection,
and if a backing Connection *is* found, use it to provide the default
arguments not explicitly passed into the hook.

boittega commented on airflow/providers/apache/spark/hooks/spark_sql.py:
```diff
@@ -72,13 +77,33 @@ def __init__(
+
         self._sql = sql
         self._conf = conf
         self._conn = self.get_connection(conn_id)
```
receive the `conn` variable here.
```suggestion
        self._conn = conn
```

uranusjr replied:
Why? I think most hooks accept an ID instead of the connection itself.

boittega replied:
you created the variable `conn` on line 89 with the same result, but inside of the try.
The idea is to only call the function `get_connection` once but you are calling it on line 89 and 109, line 109 will raise an error because it is not in a try.
Does it make sense?

uranusjr replied:
Oh, I see what you mean now. Yeah I totally missed that, thanks.

uranusjr replied:
Actually since `self._conn` is not actually used anywhere, I’ll just remove that line altogether.

boittega commented on docs/apache-airflow-providers-apache-spark/connections/spark.rst:
```diff
@@ -27,7 +27,7 @@ The Apache Spark connection type enables connection to Apache Spark.
 ----------------------
 
-Spark Submit and Spark JDBC hooks and operators use ``spark_default`` by default, Spark SQL hooks and operators point to ``spark_sql_default`` by default, but don't use it.
+Spark Submit and Spark JDBC hooks and operators use ``spark_default`` by default, Spark SQL hooks and operators point to ``spark_sql_default`` by default, but its usage is optional.
```
change to `use spark_sql_default` instead of `point to spark_sql_default`.
I would completely remove the `, but its usage is optional`, I think it is the normal behaviour for any connection.
```suggestion
Spark Submit and Spark JDBC hooks and operators use ``spark_default`` by default, Spark SQL hooks and operators use ``spark_sql_default`` by default.
```

uranusjr replied:
> I would completely remove the `, but its usage is optional`, I think it is the normal behaviour for any connection.

Is it? IIRC the Connection is *not* optional for most hooks; they fail with `AirflowNotFoundException` if a matching Connection is not found.

boittega replied:
On both Spark hooks, you have the option to send all the parameters and not using the connection as far as I know. Correct me if I am wrong.

uranusjr replied:
Hmm yeah `SparkSubmitHook` seems to also follow this approach. I guess that’s an expected approach then!

boittega commented on airflow/providers/apache/spark/hooks/spark_sql.py:
```diff
@@ -72,13 +77,33 @@ def __init__(
+                master = f"{conn.host}:{conn.port}"
+            else:
+                master = conn.host
+        if yarn_queue is None:
+            yarn_queue = options.get("queue", "default")
```
On the `self._yarn_queue = yarn_queue` bellow you can add the above logic like this:
```
self._yarn_queue = yarn_queue or options.get("queue", "default")
````

uranusjr replied:
Personally I prefer explcitly statements instead of `or`. I guess that’s subjective nad don’t have a strong opinion if there’s an established coding style. Is there?

boittega replied:
I also don't know if there is any coding style. My opinion is fewer lines and fewer variable changes.
I am happy if you want to keep it like that.

ashb reviewed (changes requested):
I'm not sure this lets us set the "spark connection type" form a connection -- i.e. `spark://master:7707`or `k8s://https://<k8s-apiserver-host>:<k8s-apiserver-port>` (see https://spark.apache.org/docs/latest/running-on-kubernetes.html#cluster-mode)

uranusjr committed:
Properly clean connections for Spark tests

uranusjr committed:
Expected Connection values in SparkSqlHook tests

Now that SparkSqlHook defaults to read values from Connection if a
config is not explicitly provided, we need to tweak the tests to reflect
this expectation.
